[FEATURE] NumPy 2.0 support

### Problem description

NumPy 2.0 is coming out soon ( https://github.com/numpy/numpy/issues/24300 ). NumPy 2.0.0rc1 packages for conda & wheels came out 3 weeks back ( https://github.com/numpy/numpy/issues/24300#issuecomment-2030603395 )

### Feature description

To prepare for NumPy 2.0, it might be worthwhile to start testing Bokeh against NumPy 2 in CI

Also as NumPy is tracking ecosystem support for NumPy 2.0, it would be helpful to share Bokeh's current support status in issue (with any plans): https://github.com/numpy/numpy/issues/26191

NumPy has put out [a migration guide]( https://numpy.org/devdocs/numpy_2_0_migration_guide.html#numpy-2-migration-guide ). More details are in [the release notes]( https://numpy.org/devdocs/release/2.0.0-notes.html ). As Bokeh doesn't have C/C++ usage of NumPy, only the Python changes would be relevant

### Potential alternatives

In the short term, Bokeh may wish to set an upper bound on NumPy version until testing can happen. Alternatively Bokeh could wait for user issues and address as those come up

Maybe there are other options that haven't been considered here

### Additional information

It appears Bokeh [depends on ContourPy]( https://github.com/bokeh/bokeh/blob/6caa4d691721b8310983c155b588a9fb94bc7824/pyproject.toml#L13 ), which may also need updates. Filed upstream issue: https://github.com/contourpy/contourpy/issues/378

There only seem to be a handful of relevant unit test failures
```
FAILED tests/unit/bokeh/core/property/test_validation__property.py::TestValidateDetailDefault::test_Bool - AssertionError: assert False
FAILED tests/unit/bokeh/plotting/test_contour.py::Test_contour_data::test_xy_dim[0] - ValueError: operands could not be broadcast together with shapes (1,1,0) (7,)
FAILED tests/unit/bokeh/plotting/test_contour.py::Test_contour_data::test_xy_dim[1] - ValueError: operands could not be broadcast together with shapes (1,1,0) (7,)
FAILED tests/unit/bokeh/plotting/test_contour.py::Test_contour_data::test_xy_dim[2] - ValueError: operands could not be broadcast together with shapes (1,1,0) (7,)
FAILED tests/unit/bokeh/plotting/test_contour.py::Test_from_contour::test_fill_line[True-True] - ValueError: operands could not be broadcast together with shapes (1,1,0) (7,)
FAILED tests/unit/bokeh/plotting/test_contour.py::Test_from_contour::test_fill_line[True-False] - ValueError: operands could not be broadcast together with shapes (1,1,0) (7,)
```

@ianthomas23 apparently so, I was testing things on two different machines, and only one of them had `contourpy` updated it seems. Contour examples and tests work fine now.

As of now I only see one failing unit test (on OSX) with 2.0.0rc1, and it *does* pass with 1.26 installed 
```
==================================================== FAILURES ====================================================
______________________________________ TestValidateDetailDefault.test_Bool _______________________________________

self = <test_validation__property.TestValidateDetailDefault object at 0x104f3fce0>

    def test_Bool(self) -> None:
        p = Bool()
        with pytest.raises(ValueError) as e:
            p.validate("junk")
>       assert matches(str(e.value), r"expected a value of type bool or bool_, got junk of type str")
E       AssertionError: assert False
E        +  where False = matches('expected a value of type bool or bool, got junk of type str', 'expected a value of type bool or bool_, got junk of type str')
E        +    where 'expected a value of type bool or bool, got junk of type str' = str(ValueError('expected a value of type bool or bool, got junk of type str'))
E        +      where ValueError('expected a value of type bool or bool, got junk of type str') = <ExceptionInfo ValueError('expected a value of type bool or bool, got junk of type str') tblen=2>.value

tests/unit/bokeh/core/property/test_validation__property.py:134: AssertionError
============================================ short test summary info =============================================
FAILED tests/unit/bokeh/core/property/test_validation__property.py::TestValidateDetailDefault::test_Bool - AssertionError: assert False
========================================= 1 failed, 111 passed in 0.33s ==========================================
```

Switching to use the `match` param of `pytest.raises` (we should update all the tests this way) provides the connection to numpy for this test:
```
    def test_Bool(self) -> None:
        p = Bool()
>       with pytest.raises(ValueError, match=r"expected a value of type bool or bool_, got junk of type str"):
E       AssertionError: Regex pattern did not match.
E        Regex: 'expected a value of type bool or bool_, got junk of type str'
E        Input: 'expected a value of type bool or bool, got junk of type str'
```

I will try to stand up a CI job with 2.0.0rc if this is the only failure then we should cut a 3.4.2 soon.

Edit: actually if this is the *only* failure, then I think we could claim that released 3.4.1 already supports numpy 2.0  since this is just the text of the error message and not the exception.

--- a/examples/server/app/selection_histogram.py
+++ b/examples/server/app/selection_histogram.py
@@ -85,7 +85,7 @@ def update(attr, old, new):
         hhist1, hhist2 = hzeros, hzeros
         vhist1, vhist2 = vzeros, vzeros
     else:
-        neg_inds = np.ones_like(x, dtype=np.bool)
+        neg_inds = np.ones_like(x, dtype=bool)
         neg_inds[inds] = False
         hhist1, _ = np.histogram(x[inds], bins=hedges)
         vhist1, _ = np.histogram(y[inds], bins=vedges)

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -145,7 +145,7 @@ lint.exclude = [
     'node_modules',
 ]
 lint.per-file-ignores = {"__init__.py" = ["F403"]}
-lint.select = ["B", "COM", "E", "F", "RUF", "TID", "UP", "W"]
+lint.select = ["B", "COM", "E", "F", "NPY", "RUF", "TID", "UP", "W"]
 lint.ignore = [
     'B005', # Using .strip() with multi-character strings is misleading the reader
     'B006', # Do not use mutable data structures for argument defaults
@@ -167,6 +167,7 @@ lint.ignore = [
     'E702', # Multiple statements on one line (semicolon)
     'E731', # Do not assign a lambda expression, use a def
     'E741', # Ambiguous variable name: I
+    'NPY002', # Replace legacy np.random.{method_name} call with np.random.Generator
     'UP035',  # Import from `collections.abc` instead: `Iterator`
     'RUF012', # Mutable class attributes should be annotated with `typing.ClassVar`
     'TID252', # Prefer absolute imports over relative imports from parent modules

